138675877 Add Account#memberships method

tmikeschu opened the pull request:
Worked with @kheppenstall and @damwhit

Add Account.find(account_id) and Account#memberships methods for related
API endpoints. All tests pass.

In order to get an account ID, we had to make a test account (which I own). We can plan on using this for the `/accounts/:id` endpoints.

[#138675877]

tmikeschu committed:
Add Account#memberships method

Add Account.find(account_id) and Account#memberships methods for related
API endpoints. All tests pass.

[#138675877]

jasonnoble commented on lib/pivotal-tracker/account.rb:
```diff
@@ -0,0 +1,30 @@
+class PivotalTracker::Account
```
```
module PivotalTracker
  class Account
```

tmikeschu replied:
@jasonnoble if we want to change this convention, we need a story for the other models as well. Also, it would still be a PivotalTracker _class_, right?

```
class PivotalTracker
  class Account
```

jasonnoble replied:
We decided to go with `class PivotalTracker::Account` because `PivotalTracker` is a class, so it can't be a module.

jasonnoble commented on lib/pivotal-tracker/account.rb:
```diff
@@ -0,0 +1,30 @@
+        PivotalTracker::AccountMembership.new(membership)
+      end
+    else
+      puts response.inspect
```
We should probably analyze the error message a bit...  It'd be cool to see "Only Admins and Owners can see memberships" or something along those lines.

tmikeschu replied:
👍

jasonnoble commented on lib/pivotal-tracker/account_membership.rb:
```diff
@@ -0,0 +1,4 @@
+class PivotalTracker::AccountMembership
+  def initialize(membership_attributes)
```
Do we need to set the values?  Similar to https://github.com/jasonnoble/pivotal-tracker/pull/5/files#diff-7d3104964801ebf75a831f703baa07a0R7 or is that another story?

tmikeschu replied:
@jasonnoble Since we have the functionality in the story working, we decided to leave this for another story.

jasonnoble replied:
Did you write up that new story?

tmikeschu replied:
Done

jasonnoble commented on lib/pivotal_tracker.rb:
```diff
@@ -7,9 +7,11 @@ class PivotalTracker
 require 'pivotal-tracker/client'
 require 'pivotal-tracker/story'
 require 'pivotal-tracker/project'
+require 'pivotal-tracker/account'
```
Should we use autoload here?  http://www.rubyinside.com/ruby-techniques-revealed-autoload-1652.html

tmikeschu replied:
@jasonnoble 👍

jasonnoble commented on spec/lib/pivotal-tracker/account_spec.rb:
```diff
@@ -0,0 +1,52 @@
+  describe "#find" do
+    let(:account) { subject.find(account_id) }
+    context "with a valid account id" do
+      let(:account_id) { 962753 }
```
Should this be a `ENV` value?

tmikeschu replied:
@jasonnoble 👍

jasonnoble commented on spec/lib/pivotal-tracker/account_spec.rb:
```diff
@@ -0,0 +1,52 @@
+
+    it "all returned objects are memberships" do
+      account.memberships.each do |membership|
+        expect(membership).to be_a(PivotalTracker::AccountMembership)
```
👍

tmikeschu committed:
Change requirements to autoloads

Changes sub-model requirements to only load when called.

[#138675877]

tmikeschu committed:
Refactor nesting syntax and create account_id ENV.

Change :: spacing to nested classes.
Remove hard-coded account_id numbers from spec.

[#138675877]

jasonnoble commented on lib/pivotal-tracker/account.rb:
```diff
@@ -0,0 +1,32 @@
+class PivotalTracker
```
`module PivotalTracker`

tmikeschu replied:
Addressed. Thanks for talking through this with me.

jasonnoble commented on lib/pivotal-tracker/account.rb:
```diff
@@ -0,0 +1,32 @@
+          PivotalTracker::AccountMembership.new(membership)
+        end
+      else
+        puts "Only Admins and Owners can see account memberships"
```
raise(PivotalTracker::PermissionDenied, "Only Admins and Owners can see account memberships")

jasonnoble replied:
In `lib/pivotal_tracker.rb`:

```
class PivotalTracker
  PermissionDenied = Class.new(Error)
end
```

tmikeschu replied:
👍

tmikeschu committed:
Add PivotalTracker::PermissionDenied error

Refactor Account model back to :: structure.

[#138675877]

tmikeschu committed:
Resolve rebase conflicts

[#138675877]
